src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -53,66 +53,226 @@ static ucs_status_t uct_rdmacm_cm_query(uct_cm_h cm, uct_cm_attr_t *cm_attr)
+    ucs_status_t               status;
+
+    cep = event->id->context;
+    ucs_assert(event->id == cep->id);
+
+    hdr = (uct_rdmacm_priv_data_hdr_t *)event->param.conn.private_data;
+
+    remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    status = uct_rdmacm_cm_id_to_dev_addr(event->id, &dev_addr, &addr_length);
+    if (status != UCS_OK) {
+        ucs_error("client (ep=%p id=%p) failed to process a connect response "
+                  "from server %s.", cep, event->id,
                   ucs_sockaddr_str(remote_addr, ip_port_str,
                                    UCS_SOCKADDR_STRING_LEN));
+        uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, status);
+        return;
+    }
+
+    remote_data.field_mask       |= UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR |
+                                    UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH;
+    remote_data.dev_addr          = dev_addr;
+    remote_data.dev_addr_length   = addr_length;
 
+    uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, hdr->status);
+
+    ucs_free(dev_addr);
+
+    if (rdma_establish(event->id)) {
+        ucs_error("rdma_establish on ep %p (to server addr=%s) failed: %m",
+                  cep, ucs_sockaddr_str(remote_addr, ip_port_str,
+                                        UCS_SOCKADDR_STRING_LEN));
+        uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, UCS_ERR_IO_ERROR);
     }
 }
 
+static void uct_rdmacm_cm_handle_event_rejected(struct rdma_cm_event *event)
+{
+    uct_rdmacm_cm_ep_t         *cep = event->id->context;
+    uct_rdmacm_priv_data_hdr_t *hdr;
+    uct_cm_remote_data_t       remote_data;
+
+    ucs_assert(event->id == cep->id);
+
+    /* Network reject or called by rdma_reject on the server */
+    remote_data.field_mask = 0;
+    if (event->param.conn.private_data != NULL) {
+        /* Server side called rdma_reject */
+        hdr = (uct_rdmacm_priv_data_hdr_t *)event->param.conn.private_data;
+        if (hdr->length > 0) {
+            remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA |
+                                                UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+            remote_data.conn_priv_data        = hdr + 1;
+            remote_data.conn_priv_data_length = hdr->length;
+
+            ucs_debug("rdmacm event rejected on ep %p (cm_id %p) with status %s.",
+                      cep, event->id, ucs_status_string(hdr->status));
+        }
+    }
+
+    uct_rdmacm_cm_ep_client_connect_cb(cep, &remote_data, UCS_ERR_REJECTED);
+}
+
+static void uct_rdmacm_cm_handle_event_established(struct rdma_cm_event *event)
+{
+    uct_rdmacm_cm_ep_t *cep = event->id->context;
+
+    ucs_assert(event->id == cep->id);
+    uct_rdmacm_cm_ep_server_connect_cb(cep, UCS_OK);
```

Comment:
i see uct_rdmacm_cm_ep_server_connect_cb() is used only once. do we really need a wrapper for it?